Proxy Configuration and Connection Leak using HTTP Clients

steph-ieffam opened the pull request:
## References
* Fixes https://github.com/DSpace/DSpace/issues/8603

## Description
Created a DSpaceHttpClientFactory to provide a client preconfigured with Proxy setting

## Instructions for Reviewers
Test network requests work as expected. This is just a refactoring


## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome).
However, reviewers may request that you complete any actions in this list if you have not done so. If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [x] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR **passes Checkstyle** validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR **includes Javadoc** for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR **passes all tests and includes new/updated Unit or Integration Tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

eliosbz committed:
[DURACOM-109] Configured proxy settings for all clients

eliosbz committed:
[DURACOM-109] Continued configuring proxy for other classes

eliosbz committed:
[DURACOM-109] Minor fix

eliosbz committed:
[DURACOM-109] Fixed http connection leaks

eliosbz committed:
[DURACOM-109] Minor fix

eliosbz committed:
[DURACOM-109] Continued fixing http connection leaks

eliosbz committed:
[DURACOM-109] Linter error fix

eliosbz committed:
[DURACOM-109] Minor fix

steph-ieffam committed:
Merge remote-tracking branch 'dspaceCommunity/main' into task/main/DURACOM-109

steph-ieffam committed:
post merge fixes

steph-ieffam committed:
fix typo

tdonohue reviewed (changes requested):
@steph-ieffam : I gave this a review & test today.  The code looks good overall (with one minor note inline below).  I tried to test as many of the modified connectors as I could.  Most work, but some are throwing errors:

1.  It's no longer possible to import Person entities via ORCID.  From MyDSpace, import a new Person Entity. Selection "ORCID" and search for "Donohue".  An error will occur.
2. Similarly, it's no longer possible to import a new Publication via ORCID.  An error will also occur on the backend.

Everything else that I've tested is still working well.  I can login via ORCID. I can import from Sherpa Journals, DataCite, OpenAIRE & other import services.  The modified Curation Tasks still appear to work. Google Captcha is still working, as is Google Analytics & Solr Stats.  Finally CC Licenses are still working.  I'm not able to easily test everything, but that's most of the modified code, so it makes me feel this is working overall.

steph-ieffam committed:
[DURACOM-109] fix typo and correct logic for ORCID connector

steph-ieffam commented on dspace-api/src/main/java/org/dspace/external/provider/impl/OrcidV3AuthorDataProvider.java:
```diff
@@ -76,37 +78,47 @@ public OrcidV3AuthorDataProvider() {
+            String authenticationParameters = "?client_id=" + clientId +
+                    "&client_secret=" + clientSecret +
+                    "&scope=/read-public&grant_type=client_credentials";
+            HttpPost httpPost = new HttpPost(OAUTHUrl + authenticationParameters);
```
thanks a lot @tdonohue for spotting these out. the above `fontbox` dependency and these changes in the ORCID class are not related to this feature. I have removed the `fontbox` dependency and restored the previous `OrcidV3AuthorDataProvider` logic which is indeed correct. The needed changes for this ORCID feature to work can be found in `dspace-api/src/main/java/org/dspace/orcid/model/factory/OrcidFactoryUtils.java`.

steph-ieffam committed:
[DURACOM-109] Orcid connector fix and improvement

steph-ieffam commented:
Hey @tdonohue thanks a lot for your review work. I addressed all the points but there's still something I wish to suggest. 
Could it be helpful to edit the `pull_request_template.md` in order to integrate a new checklist entry that says something like this:
`- [ ] If my PR performs any Network request I have made sure to use the provided DSpaceHttpClientFactory class to create the Network Client.`

I think there could be forgetfulness about using this Class as an entry point for generating the HttpClient and automatically configure the proxies. What do you think?

tdonohue reviewed (changes requested):
@steph-ieffam : I've re-reviewed and re-tested this today.  It's working *better*, but it still seems to have broken import of Person Entities via ORCID.

Here's what I'm trying.  On sandbox.dspace.org, this **works**.  But, with your PR, this **fails**:

1. Login as an Admin
2. Go to MyDSpace, select the Import button & select "Person"
3. On the import screen, select "ORCID"
4. In the Textbox, enter "Donohue" and click Search --> FAILS with this PR, but works on sandbox.dspace.org
5. In the Textbox, enter "0000-0002-9981-2089" and click Search --> FAILS with this PR, but works on sandbox.dspace.org

Everything else is working now though.  I can also verify it *is* possible to import a Publication Entity via ORCID (same process as above, and searching for "0000-0002-9981-2089" will list all publications related to that ORCID).

Finally, regarding your request related to the `pull_request_template.md`.  I don't think we should be adding specific code examples into that checkbox list, as it will eventually cause the list of checkboxes to be unwieldy.  Instead, we should look into whether there are ways to use Checkstyle (or similar code checkers) to automatically scan for improper code and throw an error or warning.  That'd be similar to how we use "lint" on the frontend to enforce strict coding standards.

steph-ieffam committed:
[DURACOM-109] added checkstyle rules to forbid usage of HttpClientBuilder.create()

steph-ieffam committed:
[DURACOM-109] fix TruncatedChunkException error

steph-ieffam commented:
Hey @tdonohue thanks for your support. I know managed to reproduce and resolve the error you mentioned. The problem was that an InputStream was being used after the connection was closed by try-with-resources block (this wasn't happening before) and the since the InputStream could not read any longer from the connection (next bytes) an error of `TruncatedChunkException` was being thrown. I know managed to solve this issue as you can see in the screenshots.

Regarding the Checkstyle validation rules I managed to add some and also added an exclusion for DSpaceHttpClientFactory since this is the entry point and we need to use the plain/base HttpClient in this java class.

Let me know if there's still something to check in this PR and I'll try to address it as soon as possible. 

![Screenshot from 2025-04-29 09-06-50](https://github.com/user-attachments/assets/b4dc8af8-9edf-4514-bd73-a509f382107e)
![Screenshot from 2025-04-29 09-06-23](https://github.com/user-attachments/assets/0fb16ea1-4e97-4348-9eb8-9e1d61f30838)

tdonohue reviewed (approved):
👍 Thanks @steph-ieffam ! Tested and reviewed this again today.  Everything is working well now, and the new checkstyle rules look good to me.  I've also double checked that the CodeQL complaint about this PR is unrelated code -- it's something that already exists in `main`.

tdonohue commented:
@steph-ieffam : This could not be automatically backported.  If you would like to see this also fixed in 8.x and 7.6.x, please created manual backport PRs.

steph-ieffam commented:
@tdonohue I have added an additional page for the proxy configuration containing the base information under the System Administration page. Mostly because I found it was hard to retrieve these configurations from the documentation without a specific page and second because I feel we should have a reference in that section to provide an easy guidance while configuring/administering DSpace
https://wiki.lyrasis.org/display/DSDOC9x/Proxy+Configuration 

Let me know if it requires changes. I have only kept track of the most important properties and added a reference to the section of the documentation where the full properties are listed
